pkg/cmd/roachtest-selector/sfselector/selector.go

```diff
@@ -0,0 +1,298 @@
+}
+
+// validate validates the SelectTestsReq
+func (r *SelectTestsReq) validate() error {
+	if _, ok := clouds[r.Cloud]; !ok {
+		keys := make([]string, 0)
+		for c := range clouds {
+			keys = append(keys, c)
+		}
+		return fmt.Errorf("unsupported cloud %s! Supported clouds: %v", r.Cloud, keys)
+	}
+	if _, ok := suites[r.Suite]; !ok {
+		return fmt.Errorf("unsupported suite %s", r.Suite)
+	}
+	return nil
+}
+
+// testDetails contains the results fetched from the query
+type testDetails struct {
+	columnHeaders []string
+	testInfos     [][]string // the fetched data
+}
+
+// SelectTests is responsible for:
+// 1. fetching the data from snowflake using the criteria provided
+// 2. uploading the results in CSV format to cloud storage
+func SelectTests(ctx context.Context, req *SelectTestsReq) error {
+	fmt.Printf("getting test details.\n")
+	r, err := getTestDetails(ctx, req)
+	if err != nil {
+		return err
+	}
+	csvBuffer, err := getCSVInMemory(r)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("uploading %s cloud test details to the cloud...\n", req.Cloud)
+	if req.DryRun {
+		fmt.Println(csvBuffer.String())
+		return nil
+	}
+	return uploadCSVToGCS(csvBuffer, req.Cloud, req.Suite)
+}
+
+// getTestDetails gets the test details from snowflake
+func getTestDetails(ctx context.Context, req *SelectTestsReq) (*testDetails, error) {
+	if err := req.validate(); err != nil {
+		return nil, err
+	}
+	db, err := getConnect(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = db.Close() }()
+	statement, err := db.Prepare(preparedQuery)
+	if err != nil {
+		return nil, err
+	}
+	// add the parameters in sequence
+	rows, err := statement.QueryContext(context.Background(),
+		req.ForPastDays*-1, fmt.Sprintf("%%%s - %s%%", suites[req.Suite], req.Cloud),
+		req.FirstRunOn*-1, req.LastRunOn*-1)
+	if err != nil {
+		return nil, err
+	}
+	// All the column headers
+	colHeaders, err := rows.Columns()
+	if err != nil {
+		return nil, err
+	}
+	// Will be used to read data while iterating rows.
+	colPointers := make([]interface{}, len(colHeaders))
+	colContainer := make([]string, len(colHeaders))
+	for i := range colPointers {
+		colPointers[i] = &colContainer[i]
+	}
+	// skipped tests are maintained separately
+	// this helps in considering them for running based on selectFromSuccessPct
+	skippedTests := make([][]string, 0)
+	td := &testDetails{
+		columnHeaders: colHeaders,
+		testInfos:     make([][]string, 0),
+	}
+	selectedTestCount := 0 // used for printing the number of tests selected
+	for rows.Next() {
+		_ = rows.Scan(colPointers...)
+		testInfos := make([]string, len(colContainer))
+		copy(testInfos, colContainer)
+		if testInfos[1] == "no" {
+			// skipped based on query
+			skippedTests = append(skippedTests, testInfos)
+		} else {
+			// selected for running
+			td.testInfos = append(td.testInfos, testInfos)
+			selectedTestCount++
+		}
+	}
+	if req.SelectFromSuccessPct > 0 && len(skippedTests) > 0 {
```

Comment:
Why does this selection happen during CSV file generation time?

It makes running a custom build, with test selection enabled, with a different "select success percentage" very difficult.

This also applies to the other parameters.

Reply:
I will make the entire run to be called from roachtest. That way, we need not worry about this.